### PR TITLE
WIP: Alpha fixes

### PR DESCRIPTION
**Pull request by bghgary**
Just a test for now. We should not merge this until there is adequate testing for the edge cases.

This will supersede the changes in #15504.

### CODE REVIEW

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Snapshot stored with reference name:
refs/pull/15573/merge

Test environment: 
https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15573/merge/index.html

To test a playground add it to the URL, for example:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15573/merge/index.html#WGZLGJ#4600

Links to test babylon tools with this snapshot:

https://playground.babylonjs.com/?snapshot=refs/pull/15573/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/15573/merge
https://gui.babylonjs.com/?snapshot=refs/pull/15573/merge
https://nme.babylonjs.com/?snapshot=refs/pull/15573/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/15573/merge#BCU1XR#0

**Comment by bjsplat** [comment]
Visualization tests for WebGL 1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15573/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

**Comment by bjsplat** [comment]
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15573/merge/testResults/webgl2playwright/index.html

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15573/merge/testResults/webgpuplaywright/index.html

File: packages/dev/core/src/Materials/PBR/pbrBaseMaterial.ts [hunk 1]
```diff
@@ -1008,55 +1002,38 @@ export abstract class PBRBaseMaterial extends PushMaterial {
...
-    protected _shouldUseAlphaFromAlbedoTexture(): boolean {
-        return this._albedoTexture != null && this._albedoTexture.hasAlpha && this._useAlphaFromAlbedoTexture && this._transparencyMode !== PBRBaseMaterial.PBRMATERIAL_OPAQUE;
-    }
-
-    /**
-     * @returns whether or not there is a usable alpha channel for transparency.
-     */
-    protected _hasAlphaChannel(): boolean {
-        return (this._albedoTexture != null && this._albedoTexture.hasAlpha) || this._opacityTexture != null;
+        return super.needAlphaTesting();
```
Comment (Popov72) [inline packages/dev/core/src/Materials/PBR/pbrBaseMaterial.ts:1036]: The `this._hasAlphaChannel()` part is missing.

File: packages/dev/core/src/Materials/standardMaterial.ts [hunk 2]
```diff
@@ -871,47 +871,17 @@ export class StandardMaterial extends PushMaterial {
...
-     */
-    protected _hasAlphaChannel(): boolean {
-        return (this._diffuseTexture != null && this._diffuseTexture.hasAlpha) || this._opacityTexture != null;
+    protected override _hasAlpha(mesh: AbstractMesh): boolean {
+        return super._hasAlpha(mesh) || this._shouldUseAlphaFromDiffuseTexture() || this._opacityTexture != null;
+    }
+
+    /** @override */
+    public override needAlphaBlending(): boolean {
+        return super.needAlphaBlending() || this._opacityFresnelParameters?.isEnabled;
```
Comment (Popov72) [inline packages/dev/core/src/Materials/standardMaterial.ts:884]: As for the PBR material, the checks on `this._opacityTexture` and `this._shouldUseAlphaFromDiffuseTexture()` are missing, and `needAlphaTesting` would also need to be overriden to check `this._hasAlphaChannel()`.

File: packages/dev/core/src/Shaders/ShadersInclude/pbrBlockAlbedoOpacity.fx [hunk 3]
```diff
@@ -81,7 +81,7 @@ albedoOpacityOutParams albedoOpacityBlock(
     #if !defined(SS_LINKREFRACTIONTOTRANSPARENCY) && !defined(ALPHAFRESNEL)
         #ifdef ALPHATEST 
             #if DEBUGMODE != 88
-                if (alpha < ALPHATESTVALUE)
+                if (alpha + 0.0000001 < ALPHATESTVALUE)
```
Comment (Popov72) [inline packages/dev/core/src/Shaders/ShadersInclude/pbrBlockAlbedoOpacity.fx:84]: For my understanding, why do we need to add this offset?

**Review by Popov72 — COMMENTED** [review]
One change is missing: in `shadowGenerator.ts`, a call to `needAlphaBlending` should be replaced with a call to `needAlphaBlendingForMesh`.

File: packages/dev/core/src/Shaders/ShadersInclude/pbrBlockAlbedoOpacity.fx (hunk 3 above)
Comment (bghgary) [reply packages/dev/core/src/Shaders/ShadersInclude/pbrBlockAlbedoOpacity.fx]: It's needed to address a quirk in the way shaders work, but we are not planning on fixing it. I've reverted this.

**Comment by sebavan** [comment]
@bghgary any update ?

**Comment by bghgary** [comment]
Not yet. On my list.

**Comment by bghgary** [comment]
Superceded by #16144.
